### ISSUE TITLE
Launchpad Pro Mk3 Various Bugs on Bitwig 3.2.7

### ISSUE THREAD

**Issue by brunobeltran** [git-moss/DrivenByMoss#203]
## Summary

After installing this extension, none of the buttons on my Launchpad Pro MK3 seem to control Bitwig Studio at all. Instead, it appears to largely function as it would if I added it as a generic midi keyboard, except that Bitwig appears to not receive input from certain pads (which pads changes with which mode but might correlate to a particular midi cc value?) 

## What I've tried

I copied the `.bwextension` file for DrivenByMoss v10.6.4 I got [here](http://www.mossgrabers.de/Software/Bitwig/Bitwig.html#3.2) into `~/Bitwig Studio/Extensions`, and the extension automatically detects and adds the Pro to my list of controllers.

Removing the extension verifies that these pads normally do work.

## System info

Running Arch Linux, `cat /proc/version` says:

`Linux version 5.8.7-arch1-1 (linux@archlinux) (gcc (GCC) 10.2.0, GNU ld (GNU Binutils) 2.35) #1 SMP PREEMPT Sat, 05 Sep 2020 12:31:32 +0000`

Bitwig's "about" says "Version: 3.2.7, Extension API Version: 12, Build date: 2020-08-11".

Let me know if there's anything I can do to help diagnose the issue! (I'm a MIDI/Bitwig neophyte, but will start trying to read the dev docs).

**Comment by brunobeltran** [comment]
In the controller script console, pressing one of the notes that "doesn't work" with the extension installed results in:

```
Mini Note on ## should be handled in framework...
Mini Note off ## should be handled in framework...
```

I'll try to keep updating this comment with my progress.

**Comment by git-moss** [comment]
Make sure that you have selected the correct MIDI in-/out. The order and naming might be different on your system. So, try all pairs till it works :-)
Also make sure that the device is set to MIDI channel 1.

**Comment by brunobeltran** [comment]
So I tried every combination of MIDI in-/out- (the novation launchpad mk3 has options 1, 2 and 3 for each). The extension defaults to MIDI in-1, as specified in [the docs](https://github.com/git-moss/DrivenByMoss-Documentation/blob/master/Novation/Novation-Launchpad.md#installation). The other two values (MIDI in-/out-2 and in-/out-3) do not appear to transmit any note values if I select them. In fact, whenever I select MIDI in-2 or MIDI in-3, the extension appears to automatically create a new controller entry that keeps the original MIDI in-1 value. 

The controller is definitely set to output to MIDI channel 1. I *can* get notes to play, it's just that some of them fail, and none of the advertised features of the extension seem to work.

I should also note that the notes that get played in "Note" mode do not correspond to the notes that are getting pressed (even when they do work), however the error messages I noted above confirm that the CC values being sent from the controller are correct and are being received by the extension.

**Comment by brunobeltran** [comment]
Update: This bug also exists when I install Bitwig Studio and the DrivenByMoss extension on Windows 10 in a Virtualbox VM. Since I had windows open, I went ahead and downloaded Ableton and tried the Launchpad in there. Everything works fine in Ableton, confirming that the issue is specifically with Bitwig Studio/DBM.

Again, let me know if there's any other troubleshooting I can do myself!

**Comment by git-moss** [comment]
No, real idea here. Even if you run Windows on a VM the connection of your external gear is still handled by Linux. Since Bitwig only officially supports Ubuntu there could still be an issue.
Maybe you could do a video showing all the configuration settings and the device. Sometimes it is easier this way to spot a mistake. Do you have access to a "real" windows system or a Mac where you could test your Launchpad?

**Comment by brunobeltran** [comment]
Got home today, borrowed the girlfriend's work laptop (which runs Windows).

I get exactly the same behavior on Windows (updater shows I'm running the latest patch) with a fresh download of Bitwig (v3.2.7). The Launchpad works totally fine *until* I install the extension.

Is this something that you can not reproduce?

I can try to help for the next couple of weeks, but I guess when the 30days of this return policy comes up if this isn't fixed unfortunately I'll have to send it back and try a different controller....I don't really have spare cash currently to buy another controller in the meantime.

**Comment by git-moss** [comment]
Maybe something simple: have you installed the latest Firmware?

**Comment by brunobeltran** [comment]
If there is a Firmware upgrade required to get the extension to work, then this is definitely the problem. 

On the Novation site right now, there is no firmware updater for the Launchpad Pro MkIII. The only firmware updater available on their site for the Launchpad Pro seems to not be for the Mk3. Tried it anyway since you asked, but it does not recognize my Launchpad. (To clarify, all of this is happening on a vanilla Windows install now).

Booting into Setup mode, the Launchpad reports a Firmware Version of "App♦0385". Do you have a newer version than this? If so I can contact Novation in the morning, or maybe wayback machine my way to victory...

**Comment by git-moss** [comment]
Their Firmware updates are done via their Components application, which you can download from the Novation site.

**Comment by brunobeltran** [comment]
What an embarrassing resolution. When I initially got the thing, the first thing I tried to do was upgrade the firmware using the Components application, but I guess it must not run correctly from inside a virtualbox (for who knows what reason).

Thank you *so* much for your patience, I really appreciate how much time and effort has gone into this, it's an amazing tool, and I'm so excited to be using my ProMk3 on Linux now thanks to you!

tl;dr: Everything works as expected after using a native windows install to upgrade the firmware.

**Comment by git-moss** [comment]
Puh, great! I was really running out of ideas!
Will add this to the manual!
